### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by mre**
I've added a bunch of new directives for clippy and fixed the highlighted warnings,
which got triggered in the process.
Also updated the dependencies to their latest versions.
Furthermore, I've added the `log::` prefix to debug output to make the use of logging more explicit.

### CODE REVIEW

File: Cargo.toml
Comment (omarandlorraine) [inline Cargo.toml:1]: do you know why dependabot didn't update this?

File: Cargo.lock
Comment (omarandlorraine) [inline Cargo.lock:1]: IIRC we removed this file from VCS. Any idea why it's back?

File: Cargo.toml
Comment (mre) [reply Cargo.toml:1]: Good question. Maybe it got disabled by GitHub?

File: Cargo.lock
Comment (mre) [reply Cargo.lock:1]: My bad. It wasn't in the `.gitignore`. Done now.

**Comment by mre** [comment]
I've fixed most of the lints.
The remaining ones are tricky, because they affect integer wrapping, e.g. `casting `u8` to `i8` may wrap around the value`. Not sure how we should handle those. So far, we don't and maybe that's the right approach, but I don't know to be honest.

@omarandlorraine, what do you think?

**Comment by mre** [comment]
Sounds like a plan, thanks

**Comment by omarandlorraine** [comment]
I've pusshed a commit to show you the kind of thing I mean; feel free to touch it up if you see anything wrong with it. I know one "problem" is that the lint still fires on the unit tests. There's probably a better way to plumb this together

and it also leaves some casts that clippy complains about

**Comment by omarandlorraine** [comment]
We've got ` (r as i8).wrapping_sub(val as i8)` on line 1009 of src/cpu.rs. This is one thing clippy fusses over at the moment. Do you think it's worth doing `#[warn(clippy::cast_possible_wrap)]` in this function? Or rewrite the expression to use the `value_is_negative` function?

That leaves one more warning where clippy gives a suggestion, how about it?

**Comment by mre** [comment]
The one issue I have with `value_is_negative` is that we could forget to call it and then run "unchecked" code.

Perhaps we could use `try_from` instead as described [here](https://github.com/rust-lang/rust-clippy/issues/9250)? [This article](https://blog.notmet.net/2021/12/rust-number-conversion-dont-follow-the-book.../) seems to suggest the same.

I'm not sure what's the more idiomatic way to do it, but it feels like that's one way to go about it. In any case, I'm open for suggestions here. 😉 

Thanks for looking into this.

**Comment by omarandlorraine** [comment]
>Perhaps we could use try_from instead as described https://github.com/rust-lang/rust-clippy/issues/9250?

While I do agree with you, the cited articles and clippy, that `as` is not the solution we're looking for, it is defined across the whole range of `i8`, which we wouldn't get with `u8::try_from`. 

Since N and Z often get updated together, we could do a method taking a u8 called `update_nz`, taking a `u8`. We could even make it generic for `u8` and `i8` if the need arises.

**Comment by omarandlorraine** [comment]
> We could even make it generic for `u8` and `i8` if the need arises.

Actually scrap that. I think that all the emulated ALU's arithmetic works on u8; there is very little need for casting any more

**Comment by mre** [comment]
Just to clarify, are the casts in the current code an actual problem for us?
Because alternatively, we leave everything as is and disable the lint. I find the existing code quite readable; I'm just a little concerned about potential bugs.

**Comment by omarandlorraine** [comment]
> Just to clarify, are the casts in the current code an actual problem for us?
> Because alternatively, we leave everything as is and disable the lint. I find the existing code quite readable; I'm just a little concerned about potential bugs. 

I think it is a stylistic problem, and the lint is doing its job. 

But functionally, the current code is doing exactly what it should, as solid65 confirms. 

The current code can be changed to avoid the cast while still being correct, and this is what I think we should do.

**Comment by mre** [comment]
Went ahead and squashed the commits.
Added a few lint exceptions for the tests and documented the reasoning.

**Comment by mre** [comment]
@omarandlorraine, okay to merge?

**Comment by omarandlorraine** [comment]
> @omarandlorraine, okay to merge?

Okay by me! Thanks for a good job!

**Pull request merged**